s2cloudless/utils.py

```diff
@@ -54,3 +70,83 @@ def get_s2_evalscript(all_bands: bool = False, reflectance: bool = False) -> str
         metadata_evalscript="" if reflectance else METADATA_EVALSCRIPT_TEMPLATE,
     ).strip("\n ")
+
+
+def get_timestemps(
+    bbox: BBox,
+    time_interval: Tuple[dt.datetime, dt.datetime],
+    *,
+    maxcc: Optional[float] = None,
+    data_collection: DataCollection = DataCollection.SENTINEL2_L1C,
+    config: Optional[SHConfig] = None,
+    time_difference: Optional[dt.timedelta] = None,
+) -> List[dt.datetime]:
+    """Get the list of timestamps from within date range for which data of the bbox is available."""
+    time_difference = time_difference if time_difference else dt.timedelta(seconds=0)
+
+    catalog = SentinelHubCatalog(config=config)
+    cloud_cover_query = None
+    if maxcc is not None:
+        cloud_cover_query = f"eo:cloud_cover < {100 * float(maxcc)}"
+
+    search_iterator = catalog.search(
+        data_collection,
+        bbox=bbox,
+        time=time_interval,
+        filter=cloud_cover_query,
+        fields={
+            "include": [
+                "properties.datetime",
+            ],
+            "exclude": [],
+        },
+    )
+
+    return filter_times(search_iterator.get_timestamps(), time_difference)
+
+
+def download_bands_and_valid_data_mask(
+    bbox: BBox,
+    timestamps: List[dt.datetime],
+    *,
+    data_collection: DataCollection = DataCollection.SENTINEL2_L1C,
+    config: Optional[SHConfig] = None,
+    size: Optional[Tuple[int, int]] = None,
+    resolution: Optional[Tuple[float, float]] = None,
+    all_bands: bool = True,
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Downloaded specifid data"""
+
+    client = SentinelHubDownloadClient(config=config)
+
+    api_requests = []
+    for timestamp in timestamps:
+        request = SentinelHubRequest(
+            evalscript=get_s2_evalscript(all_bands=all_bands, reflectance=False),
+            input_data=[
+                SentinelHubRequest.input_data(
+                    data_collection=data_collection,
+                    time_interval=(timestamp, timestamp),
+                )
+            ],
+            responses=[
+                SentinelHubRequest.output_response("default", MimeType.TIFF),
+                SentinelHubRequest.output_response("userdata", MimeType.JSON),
+            ],
+            bbox=bbox,
+            size=size,
+            resolution=resolution,
+            config=config,
+        )
+
+        api_requests.append(request)
+
+    responses = client.download([api_request.download_list[0] for api_request in api_requests])
+
+    # pylint: disable=not-an-iterable
+    data = np.asarray([response["default.tif"] for response in responses], dtype=np.float32)
+    norm_factors = [response["userdata.json"]["norm_factor"] for response in responses]
+    del responses
+
+    normalized_bands = (np.round(array * factor, 4) for array, factor in zip(data[..., :-1], norm_factors))
+    return np.asarray(list(normalized_bands), dtype=np.float32), data[..., -1] != 0
```

Comment:
I'd do it like this:
1. extract normalization factors
2. extract and normalize data
3. transform to numpy array and output

Reply:
updated